src/backend/optimizer/plan/createplan.c

```diff
@@ -6303,6 +6357,54 @@ prepare_sort_from_pathkeys(Plan *lefttree, List *pathkeys,
 			lefttree->targetlist = tlist;	/* just in case NIL before */
 		}
 
+		/* Get the ndist value from pg_stats */
+		if (mkqsApplicable && root != NULL)
+		{
+			/*
```

Comment:
If mkqsApplicable is TRUE, root is not NULL. 
if (mkqsApplicable)
{
      Assert(root);
}
is better?

Reply:
Actually root may be NULL when `mkqsApplicable` is TRUE, at least for mergejoin plan. Please see the code path:

```
#0  0x000055e45969b30b in prepare_sort_from_pathkeys (root=0x0, lefttree=0x55e45ab24260, pathkeys=0x55e45ab228a8,
    relids=0x55e45aa8c4d8, reqColIdx=0x0, adjust_tlist_in_place=<optimized out>, p_numsortkeys=0x7ffdd428e2f4,
    p_sortColIdx=0x7ffdd428e308, p_sortOperators=0x7ffdd428e310, p_collations=0x7ffdd428e318,
    p_nullsFirst=0x7ffdd428e320, mkqsApplicable=0x7ffdd428e2f3) at createplan.c:6371
#1  0x000055e45969f330 in make_sort_from_pathkeys (relids=<optimized out>, pathkeys=<optimized out>,
    lefttree=0x55e45ab24260, root=0x0) at createplan.c:6523
#2  create_mergejoin_plan (best_path=0x55e45ab22a10, root=0x55e45aa76c50) at createplan.c:4546
#3  create_join_plan (root=0x55e45aa76c50, best_path=0x55e45ab22a10) at createplan.c:1094
#4  0x000055e45969b8fb in create_plan_recurse (root=0x55e45aa76c50, best_path=0x55e45ab22a10, flags=6)
    at createplan.c:423
#5  0x000055e45969c105 in create_incrementalsort_plan (flags=4, best_path=0x55e45ab233c0, root=0x55e45aa76c50)
    at createplan.c:2238
#6  create_plan_recurse (root=0x55e45aa76c50, best_path=0x55e45ab233c0, flags=4) at createplan.c:499
#7  0x000055e45969bebe in create_agg_plan (best_path=0x55e45ab23500, root=0x55e45aa76c50) at createplan.c:2336
#8  create_plan_recurse (root=0x55e45aa76c50, best_path=0x55e45ab23500, flags=3) at createplan.c:514
#9  0x000055e45969c63d in create_sort_plan (flags=1, best_path=0x55e45ab23b10, root=0x55e45aa76c50)
    at createplan.c:2207
#10 create_plan_recurse (root=0x55e45aa76c50, best_path=0x55e45ab23b10, flags=1) at createplan.c:494
#11 0x000055e45969c2b2 in create_limit_plan (flags=1, best_path=0x55e45ab23e78, root=0x55e45aa76c50)
    at createplan.c:2882
#12 create_plan_recurse (root=0x55e45aa76c50, best_path=0x55e45ab23e78, flags=1) at createplan.c:541
#13 0x000055e45969ead3 in create_plan (root=root@entry=0x55e45aa76c50, best_path=best_path@entry=0x55e45ab23e78)
    at createplan.c:354
#14 0x000055e4596aec2a in standard_planner (parse=0x55e45aa55158, query_string=<optimized out>,
    cursorOptions=2048, boundParams=<optimized out>) at planner.c:6338
#15 0x000055e4597a1984 in pg_plan_query (querytree=querytree@entry=0x55e45aa55158,
    query_string=query_string@entry=0x55e45a955e68 "EXPLAIN (COSTS OFF) \n    SELECT col12, count(distinct a.col1), count(distinct a.col2), count(distinct b.col1), count(distinct b.col2), count(*)\n    FROM test_mark_restore a\n        JOIN test_mark_rest"..., cursorOptions=cursorOptions@entry=2048, boundParams=boundParams@entry=0x0)
    at postgres.c:904
#16 0x000055e45951ff69 in standard_ExplainOneQuery (query=0x55e45aa55158, cursorOptions=cursorOptions@entry=2048,
    into=into@entry=0x0, es=es@entry=0x55e45aa3b050,
    queryString=0x55e45a955e68 "EXPLAIN (COSTS OFF) \n    SELECT col12, count(distinct a.col1), count(distinct a.col2), count(distinct b.col1), count(distinct b.col2), count(*)\n    FROM test_mark_restore a\n        JOIN test_mark_rest"..., params=params@entry=0x0, queryEnv=0x0) at explain.c:483
#17 0x000055e45952085c in ExplainOneQuery (queryEnv=<optimized out>, params=0x0, queryString=<optimized out>,
    es=0x55e45aa3b050, into=0x0, cursorOptions=2048, query=<optimized out>) at explain.c:439
#18 ExplainQuery (pstate=pstate@entry=0x55e45aa3af48, stmt=stmt@entry=0x55e45aa54fb8, params=params@entry=0x0,
    dest=dest@entry=0x55e45aa3aec0) at explain.c:335
#19 0x000055e4597a795e in standard_ProcessUtility (pstmt=0x55e45aa55050,
    queryString=0x55e45a955e68 "EXPLAIN (COSTS OFF) \n    SELECT col12, count(distinct a.col1), count(distinct a.col--Type <RET> for more, q to quit, c to continue without paging--f 0
2), count(distinct b.col1), count(distinct b.col2), count(*)\n    FROM test_mark_restore a\n        JOIN test_mark_rest"..., readOnlyTree=<optimized out>, context=PROCESS_UTILITY_TOPLEVEL, params=0x0, queryEnv=0x0,
    dest=0x55e45aa3aec0, qc=0x7ffdd428ef00) at utility.c:863
#20 0x000055e4597a50e5 in PortalRunUtility (portal=portal@entry=0x55e45a9dc3b8, pstmt=0x55e45aa55050,
    isTopLevel=isTopLevel@entry=true, setHoldSnapshot=setHoldSnapshot@entry=true, dest=dest@entry=0x55e45aa3aec0,
    qc=qc@entry=0x7ffdd428ef00) at pquery.c:1158
#21 0x000055e4597a5737 in FillPortalStore (portal=0x55e45a9dc3b8, isTopLevel=<optimized out>)
    at ../../../src/include/nodes/pg_list.h:281
#22 0x000055e4597a644f in PortalRun (portal=portal@entry=0x55e45a9dc3b8, count=count@entry=9223372036854775807,
    isTopLevel=isTopLevel@entry=true, run_once=run_once@entry=true, dest=dest@entry=0x55e45aa84970,
    altdest=altdest@entry=0x55e45aa84970, qc=0x7ffdd428f110) at pquery.c:763
#23 0x000055e4597a1d77 in exec_simple_query (
    query_string=0x55e45a955e68 "EXPLAIN (COSTS OFF) \n    SELECT col12, count(distinct a.col1), count(distinct a.col2), count(distinct b.col1), count(distinct b.col2), count(*)\n    FROM test_mark_restore a\n        JOIN test_mark_rest"...) at postgres.c:1274
#24 0x000055e4597a3bcb in PostgresMain (dbname=<optimized out>, username=<optimized out>) at postgres.c:4680
#25 0x000055e45979e3e3 in BackendMain (startup_data=<optimized out>, startup_data_len=<optimized out>)
    at backend_startup.c:105
#26 0x000055e4596ef4b7 in postmaster_child_launch (child_type=child_type@entry=B_BACKEND,
    startup_data=startup_data@entry=0x7ffdd428f610 "", startup_data_len=startup_data_len@entry=4,
    client_sock=client_sock@entry=0x7ffdd428f630) at launch_backend.c:265
#27 0x000055e4596f3753 in BackendStartup (client_sock=0x7ffdd428f630) at postmaster.c:3593
#28 ServerLoop () at postmaster.c:1674
#29 0x000055e4596f5fad in PostmasterMain (argc=<optimized out>, argv=0x55e45a94fbb0) at postmaster.c:1372
#30 0x000055e4593a8db0 in main (argc=3, argv=0x55e45a94fbb0) at main.c:197
```